feat(#425): E2E test actions

m5r opened the pull request:
# Description

medic/cht-conf#425

This PR adds a first E2E test covering the `TBD` action. Tests are run against an actual CHT instance using [CHT Docker helper](https://docs.communityhealthtoolkit.org/hosting/4.x/app-developer/#cht-docker-helper-for-4x) and are run automatically in the CI.

# Code review items

- Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- Tested: Unit and/or integration tests where appropriate
- Backwards compatible: Works with existing data and configuration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

mrjones-plip commented:
@m5r - instead of using an upstream branch on CHT Core (`dnm-docker-helper-experiments` which [hard codes](https://github.com/medic/cht-core/blob/dnm-docker-helper-experiments/scripts/docker-helper-4.x/cht-docker-compose.sh#L462-L463) IP in `curl` call in Docker Helper), instead lets update the runner to have an entry in `/etc/hosts` before it tries to fetch the certs from local-ip server.

Something like this should work (untested!) :

```
- name: Hard code local-ip IP in /etc/hosts per https://github.com/medic/medic-infrastructure/issues/571#issuecomment-2209120441
  run: |
      echo "15.188.129.97 local-ip.medicmobile.org" | sudo tee -a /etc/hosts
```

jkuester commented on test/e2e/cht-docker-utils.js:
```diff
@@ -0,0 +1,135 @@
+      resolve();
+    });
+
+    childProcess.stdin.write('y\n');
```
Should we maybe log a cht-core ticket to enhance the docker-helper to support taking arguments on the initial command instead of requiring interactive prompts?  This  works fine here for now, but will fail if prompts are added/removed from the docker-helper.  (Also, being able to pass everything as args would be generally useful for the docker-helper since it would make it easier to script with...)

mrjones-plip replied:
I was thinking about this to! While it'd be nice if the helper functionality was in `cht-conf` as has been kicked around, instead:

I wondered about a docker helper refactor that would break out all the functions, but specifically the `init_env_file()` [function](https://github.com/medic/cht-core/blob/e6d184946affc62773d569168216a5b913f38a30/scripts/docker-helper-4.x/cht-docker-compose.sh#L17) and `create_compose_files()`  [function](https://github.com/medic/cht-core/blob/e6d184946affc62773d569168216a5b913f38a30/scripts/docker-helper-4.x/cht-docker-compose.sh#L66C1-L66C21) into a stand alone "docker helper function" file.  These could then be included in docker helper with effectively no code changes while allowing external scripts (like this e2e test) use a loose API of sorts to legally, but programmatically call the to exact functions.  As it's a loose API, we can ensure a bit of backwards comparability going forwards.

If this sounds helpful and viable - happy to jump on this refactor!

m5r replied:
Honestly I love the idea! To push it further, while working on this setup, I was contemplating how convenient it would be to have a javascript _script_ instead of a bash script. Instead of having a bunch of `child_process.spawn('cht-docker-compose.sh')` to interface with the script, we could call javascript functions to orchestrate all this.

Technically it wouldn't change anything for app devs, instead of running `./cht-docker-compose.sh` they would run `./cht-docker-compose.js`. As long as the file has the execute permission `+x` and a shebang at the top `#!/usr/bin/env -S node`, it will work fine

For this PR specifically I don't think we should make any change but we can already open a cht-core issue to track it and discuss it there

jkuester replied:
https://github.com/medic/cht-core/issues/9261

jkuester commented on test/e2e/edit-app-settings.spec.js:
```diff
@@ -0,0 +1,69 @@
+
+const { getProjectUrl } = require('./cht-docker-utils');
+const {
+  DEFAULT_PROJECT_NAME,
```
Looks like this field does not exist in cht-conf-utils?  (Tests now are just creating a project in `build/undefined`).

m5r replied:
where is typescript when we need it... 🤦‍♂️ good catch!

jkuester commented on test/e2e/cht-conf-utils.js:
```diff
@@ -0,0 +1,57 @@
+
+const initProject = async (projectName) => {
+  const projectDirectory = getProjectDirectory(projectName);
+  if (fs.existsSync(projectDirectory)) {
```
Minor, but to reduce some duplicate code, I would just add this if-check to the `cleanupProject` function and then call that function here.

jkuester commented on test/e2e/cht-conf-utils.js:
```diff
@@ -0,0 +1,57 @@
+  }
+
+  fse.mkdirpSync(projectDirectory);
+  fs.writeFileSync(
```
What do we actually need this package.json file for?  The current test seems to run fine without it (but maybe you are anticipating something in the future).  

For the current script code, it seems like we don't need the cht-conf dependency here since `runChtConf` is directly calling the `index.js` file.

m5r replied:
I believe the only scenario where a `package.json` file would be required is if a test covers [`src/lib/check-cht-conf-dependency-version.js`](https://github.com/medic/cht-conf/blob/main/src/lib/check-cht-conf-dependency-version.js). Since it's closer to what an actual config directory looks like, I would rather keep it

jkuester commented on test/e2e/edit-app-settings.spec.js:
```diff
@@ -0,0 +1,69 @@
+
+    // TODO: remove next line when we upgrade eslint and its `parserOptions.ecmaVersion` setting to parse syntax supported by node 18+
+    // eslint-disable-next-line no-undef
+    const baseSettings = structuredClone(initialSettings);
```
What is the need of doing a structuredClone if you are never referencing `initialSettings` again?

m5r replied:
In the first iterations I was reusing `initialSettings` and that led me to clone it with `structuredClone` and I left it there after changing the code. It's not useful anymore so I'll get rid of it, nice catch

jkuester replied:
I am excited to see `structuredClone` actually getting some use! Definitely been needing that function (but I guess not in this case...)!

![shoul](https://i.kym-cdn.com/entries/icons/original/000/040/653/goldblum-quote.jpeg)

jkuester commented on test/e2e/cht-docker-utils.js:
```diff
@@ -0,0 +1,135 @@
+        .map(line => line.split('='))
+        .filter(entry => entry.length === 2),
+    );
+  } catch (error) {
```
Is there a particular workflow where you know we need this `catch`?  I am just not sure what the use is of recovering gracefully if we cannot find the config file....

m5r replied:
That's a good point, I figured we could try the default port in case the instance is somehow running but it makes more sense to throw early and let it crash in this unsupported scenario

jkuester commented on test/e2e/edit-app-settings.spec.js:
```diff
@@ -0,0 +1,69 @@
+
+  it('disables a language, recompile, and push app settings', async () => {
+    const url = await getProjectUrl(projectName);
+    const initialSettings = await request.get({ url: `${url}/api/v1/settings`, json: true });
```
Minor, but seems worth adding a few assertions about the initial state of the settings (`local`, etc) before we edit things.  (Since we are so far removed from how/where these initial settings are being defined).

jkuester commented on test/e2e/edit-app-settings.spec.js:
```diff
@@ -0,0 +1,69 @@
+    });
+    expect(newSettings.locale).to.equal('fr');
+    expect(newSettings.locale_outgoing).to.equal('fr');
+
```
```suggestion
```

jkuester commented on test/e2e/edit-app-settings.spec.js:
```diff
@@ -0,0 +1,69 @@
+    await fs.promises.writeFile(
+      path.join(projectDirectory, 'app_settings/base_settings.json'),
+      JSON.stringify(baseSettings, null, 2),
+    );
```
I think we are going to want a layer of util code for writing/reading these standard config files for a given `projectDirectory`.  Maybe these functions could go into `cht-conf-utils.js` or a different dedicated file so they can be re-used in other tests.

jkuester commented on test/e2e/cht-docker-utils.js:
```diff
@@ -0,0 +1,134 @@
+    await fs.promises.mkdir(dockerHelperDirectory);
+  }
+
+  if (!fs.existsSync(dockerHelperScript)) {
```
An alternative would be to maybe cook up some kind of npm `postinstall` script that re-downloads the script any time `npm ci` or `npm install` is run.  (Though I would want to be sure the `postinstall` only gets executed on the dev workflows and not when an app developer is just trying to globally install cht-conf.

m5r replied:
I agree with the removal of `.cht-docker-helper` at the end of the teardown but I'd rather keep those checks here because someoen could potentially cancel the test run and end up not running the `tearDownCht` function, resulting in a running cht instance and leftover files on the system.
That way we keep running `tearDownCht` in the `before` hook and clean up any leftovers

tatilepizs reviewed (approved):
Thank you for adding this test @m5r, it looks great ⭐ 

Just one thing, I didn't find a section in `cht-docs` that was related to `cht-conf`, only the [README file](https://github.com/medic/cht-conf), I don't know if this was already discussed at some point and we decided to manage the `cht-conf` documentation outside the `cht-docs` 🤷‍♀️ but I think that it will be important to add a section about these e2e tests (in the `README` file or maybe create a new section in `cht-docs`) to talk about how can we run them (`npm run test-e2e`) and maybe about the possible errors that we can find. 
For example I had the same error that Josh mention [here](https://github.com/medic/cht-conf/pull/620/files#r1671327190), luckily I read that comment before trying to execute the tests 😅 , but then I had a error because my CHT server was never up, and I had to increase the `timeout` value to `100_000` to be able to run it locally.

We can create something similar (but smaller) to the tests specification for `cht-core` -> [Automated Tests](https://docs.communityhealthtoolkit.org/contribute/code/core/automated-tests/)

m5r commented on package.json:
```diff
@@ -11,7 +11,8 @@
     "docker-stop-couchdb": "docker stop cht-conf-couchdb || true",
-    "test": "npm run eslint && npm run docker-start-couchdb && npm run clean && mkdir -p build/test && cp -r test/data build/test/data && cd build/test && nyc --reporter=html mocha --forbid-only \"../../test/**/*.spec.js\" && cd ../.. && npm run docker-stop-couchdb",
+    "test": "npm run eslint && npm run docker-start-couchdb && npm run clean && mkdir -p build/test && cp -r test/data build/test/data && cd build/test && nyc --reporter=html mocha --forbid-only \"../../test/**/*.spec.js\" --exclude \"../../test/e2e/**/*.spec.js\" && cd ../.. && npm run docker-stop-couchdb",
+    "test-e2e": "npm run eslint && mocha --config test/e2e/.mocharc.js",
```
Nope, no other reason than copy/pasting the `test` command right above. I'll get rid of it, good catch!

m5r commented on .github/workflows/build.yml:
```diff
@@ -30,3 +30,19 @@ jobs:
           .nyc_output
       if: ${{ failure() }}
+
+  e2e:
```
`build` is not an accurate name for that CI job, there is no build step for cht-conf, it would be clearer if it was named `unit-tests`
with that said, I think it makes sense to run all tests in parallel like we do in cht-core and I would prefer to keep it this way if that makes sense to you too

jkuester replied:
Ah good point!  Had not thought of it that way.  Makes sense. :+1:

m5r commented:
@tatilepizs excellent idea! First, I've added logs and increased the timeout to 120 seconds. This should be enough as it rarely takes more than 60 seconds in the CI to spin up an instance. cht-core's API tests have a 200 second timeout but I don't think we need to reach for that high timeout at this stage.
Second, I've added a section in the README to cover the `npm run e2e-test` command and how tests are run. Let me know if anything is missing/needs clarification.

And finally to answer your question, I don't know either why cht-docs doesn't have much covering cht-conf..

mrjones-plip commented:
> I don't know either why cht-docs doesn't have much covering cht-conf...

This is both a question and an answer ;)  It's not been documented because people keep asking the question!!

If either of you are interested in starting docs on this, possibly with a first step of creating a docs page and replacing the enire readme  with a "see docs site!" link, that'd be great!

cc @esthermmoturi & @nsitaula

tatilepizs commented:
😅 I can help with that. I am going to create a PR to move the README file to the docs. thanks @mrjones-plip 

And thanks @m5r for the improvements and the section in the README file 🌟

jkuester reviewed (approved):
LGTM!

garethbowen commented:
@m5r NB this should have been a `chore` not a `feat` as it's not something that a CLI user will notice.
